beetsplug/musicbrainz.py

```diff
@@ -0,0 +1,919 @@
+    date_str: str,
+    original: bool = False,
+):
+    """Given a (possibly partial) YYYY-MM-DD string and an AlbumInfo
+    object, set the object's release date fields appropriately. If
+    `original`, then set the original_year, etc., fields.
+    """
+    if date_str:
+        date_parts = date_str.split("-")
+        for key in ("year", "month", "day"):
+            if date_parts:
+                date_part = date_parts.pop(0)
+                try:
+                    date_num = int(date_part)
+                except ValueError:
+                    continue
+
+                if original:
+                    key = "original_" + key
+                setattr(info, key, date_num)
+
+
+def _parse_id(s: str) -> str | None:
+    """Search for a MusicBrainz ID in the given string and return it. If
+    no ID can be found, return None.
+    """
+    # Find the first thing that looks like a UUID/MBID.
+    match = re.search("[a-f0-9]{8}(-[a-f0-9]{4}){3}-[a-f0-9]{12}", s)
+    if match is not None:
+        return match.group() if match else None
+    return None
+
+
+def _is_translation(r):
+    _trans_key = "transl-tracklisting"
+    return r["type"] == _trans_key and r["direction"] == "backward"
+
+
+def _find_actual_release_from_pseudo_release(
+    pseudo_rel: JSONDict,
+) -> JSONDict | None:
+    try:
+        relations = pseudo_rel["release"]["release-relation-list"]
+    except KeyError:
+        return None
+
+    # currently we only support trans(liter)ation's
+    translations = [r for r in relations if _is_translation(r)]
+
+    if not translations:
+        return None
+
+    actual_id = translations[0]["target"]
+
+    return musicbrainzngs.get_release_by_id(actual_id, RELEASE_INCLUDES)
+
+
+def _merge_pseudo_and_actual_album(
+    pseudo: beets.autotag.hooks.AlbumInfo, actual: beets.autotag.hooks.AlbumInfo
+) -> beets.autotag.hooks.AlbumInfo | None:
+    """
+    Merges a pseudo release with its actual release.
+
+    This implementation is naive, it doesn't overwrite fields,
+    like status or ids.
+
+    According to the ticket PICARD-145, the main release id should be used.
+    But the ticket has been in limbo since over a decade now.
+    It also suggests the introduction of the tag `musicbrainz_pseudoreleaseid`,
+    but as of this field can't be found in any official Picard docs,
+    hence why we did not implement that for now.
+    """
+    merged = pseudo.copy()
+    from_actual = {
+        k: actual[k]
+        for k in [
+            "media",
+            "mediums",
+            "country",
+            "catalognum",
+            "year",
+            "month",
+            "day",
+            "original_year",
+            "original_month",
+            "original_day",
+            "label",
+            "barcode",
+            "asin",
+            "style",
+            "genre",
+        ]
+    }
+    merged.update(from_actual)
+    return merged
+
+
+class MusicBrainzPlugin(BeetsPlugin):
```

Comment:
Shouldn't this be a MetadataSourcePlugin? Is their a reason it isnt one?

Reply:
Good question, just had a look at the rest of internal autotagger plugins:

| source          | base                   | uses 3rd party library |
| --------------- | ---------------------- | ---------------------- |
| **spotify**     | `MetadataSourcePlugin` | no                     |
| **deezer**      | `MetadataSourcePlugin` | no                     |
| **discogs**     | `BeetsPlugin`          | yes                    |
| **musicbrainz** | `BeetsPlugin`          | yes                    |

It seems to me that `MetadataSourcePlugin` does indeed expect the plugin to interact with metadata sources directly: see `search_url` property and `_search_api` method for example.

Reply:
I see, thanks!

Should we generalize the MetadataSourcePlugin to allow for 3rd party api wrappers than? (Maybe we should create an issue for that, I see that this could be out of scope here)

Having all data source plugins with one common parent class is preferable, imo.

Reply:
> Having all data source plugins with one common parent class is preferable, imo.

Agree with you, let's create an issue for this as it's indeed a bit out of scope here

Reply:
Sure thing. I can start one, but I will need some assistance as this is more of a conceptual/system design problem and I'm not the person who should make these decisions on my own :)